gst/nnstreamer/tensor_source/tensor_src_iio.c

```diff
@@ -2090,33 +2104,51 @@ gst_tensor_src_iio_fill (GstBaseSrc * src, guint64 offset,
   g_assert (gst_memory_map (mem, &map, GST_MAP_WRITE));
 
   /** wait for the data to arrive */
-  if (self->trigger.name != NULL) {
-    status = poll (self->buffer_data_fp, 1, -1);
-    if (status < 0) {
-      GST_ERROR_OBJECT (self, "Error %d while polling the buffer.", status);
-      goto error_mem_unmap;
-    } else if (status == 0) {
-      /** this case happens for timeout, which is not set yet */
-      GST_ERROR_OBJECT (self, "Timeout while polling the buffer.");
-      goto error_mem_unmap;
+  time_to_end = g_get_real_time () + self->poll_timeout * 1000;
+  while (TRUE) {
+    if (self->trigger.name != NULL) {
+      status = poll (self->buffer_data_fp, 1, self->poll_timeout);
+      if (status < 0) {
+        GST_ERROR_OBJECT (self, "Error %d while polling the buffer.", status);
+        goto error_mem_unmap;
+      } else if (status == 0) {
+        GST_ERROR_OBJECT (self, "Timeout while polling the buffer.");
+        goto error_mem_unmap;
+      } else if (!(self->buffer_data_fp->revents & POLLIN)) {
+        GST_ERROR_OBJECT (self, "Poll succeeded on an unexpected event %d.",
+            self->buffer_data_fp->revents);
+        goto error_mem_unmap;
+      }
+      self->buffer_data_fp->revents = 0;
+    } else {
+      /** sleep for a device tick */
+      g_usleep (MAX (1,
+              (guint64) 1000000 / (self->sampling_frequency /
+                  self->buffer_capacity)));
```

Comment:
For higher precision, do multiplication first as long as there is no worries on overflow.